### ISSUE TITLE
Merge and Migrate Data from Sharded Tables,there is a problem with the causality logic, and the downstream data is inconsistent

### ISSUE THREAD

**Issue by wxiaoke** [pingcap/tiflow#9089]
### What did you do?

1. Start MySQL 5.7 and populate data
```
sudo docker run --name m5 -e MYSQL_ROOT_PASSWORD=111 -d mysql:5.7 --log-bin=ON --server-id=100 --binlog-format=ROW --show-compatibility-56=ON
mysql -u root -h 172.17.0.2 -p111
create schema test;
create table test.t1 (a bigint primary key, b varchar(256));
create table test.t2 (a bigint primary key, b varchar(256));
insert into test.t1 values (3, 'three'), (5, 'five'), (7, 'seven');
insert into test.t2 values (2, 'two'), (4, 'four'), (6, 'six');
```
2. Start TiDB downstream
```
rm -rf /tmp/tidb
tiup tidb:v6.5.0
```
3. Start DM
```
$TIFLOW_PATH/dm-master --name master1 --master-addr 127.0.0.1:8261 --peer-urls 127.0.0.1:8291 --initial-cluster 'master1=http://127.0.0.1:8291'
$TIFLOW_PATH/dm-worker --name worker1 --join 127.0.0.1:8261 --worker-addr 127.0.0.1:8262 -L debug
```

4. Create DM source and task

```
# ticket2178-src.yml
source-id: ticket2178-src
from:     host: '172.17.0.2'
    port: 3306
    user: 'root'
    password: '111'
# ticket2178-task.yml
name: ticket2178-task
task-mode: all
target-database:     host: '127.0.0.1'
    port: 4000
    user: 'root'
    password: ''
mysql-instances:   - source-id: ticket2178-src
    route-rules: ['r1']

routes:      r1:
        schema-pattern: 'test'
        table-pattern: 't*'
        target-schema: 'test'
        target-table: 't'
$TIFLOW_PATH/dmctl --master-addr 127.0.0.1:8261 operate-source create ./ticket2178-src.yml
$TIFLOW_PATH/dmctl --master-addr 127.0.0.1:8261 check-task ./ticket2178-task.yml
$TIFLOW_PATH/dmctl --master-addr 127.0.0.1:8261 start-task ./ticket2178-task.yml
$TIFLOW_PATH/dmctl --master-addr 127.0.0.1:8261 query-status ticket2178-task
mysql -u root -h 127.0.0.1 -P 4000 -e 'select * from test.t;'
```
5. Perform the DML

```
mysql -u root -h 172.17.0.2 -p111
BEGIN;
DELETE FROM test.t2 WHERE a = 2;
INSERT INTO test.t1 VALUES (2, 'two*');
COMMIT;
```
6. Check for the causality key using the keyword component=causality
![image](https://github.com/pingcap/tiflow/assets/38338933/d6714aaf-5415-49cd-a2ab-abf1710af82d)

### Versions of the cluster

DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):
```console
(paste DM version here, and you must ensure versions of dmctl, DM-worker and DM-master are same)
all LTS version will encounter this issue.
```
Upstream MySQL/MariaDB server version:
```console
(paste upstream MySQL/MariaDB server version here)
mysql 5.7/ mysql 8
actually, this is not related to MySQL version
```
Downstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):
```console
(paste TiDB cluster version here)
tidb 4.x- 6.x
actually, this is not related to tidb version
```
How did you deploy DM: tiup or manually?
```console
(leave TiUP or manually here)
```
Other interesting information (system version, hardware config, etc):
```console
>
>
```

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

**Comment by kennytm** [comment]
TL;DR: The causality key used the upstream table name, rather than the downstream table name. This leads to causality violation in shard-merging scenario when a DML with the same PK appeared in different shards, because a conflict cannot be detected.

**Comment by lance6716** [comment]
This is not supported. Even if we use downstream table name in causality key, it can't handle the case where different shard tables are bind to different DM workers different nodes. I think you can try to use `routes.${route-rule}.extract-table/extract-schema/extract-source` to extract information to extra columns, and make a composite primary key with these columns.

https://docs.pingcap.com/tidb/stable/task-configuration-file-full

**Comment by kennytm** [comment]
@lance6716 OK. I suppose setting `worker-count: 1` can also restrict the entire processing to a single worker node?

**Comment by lance6716** [comment]
> @lance6716 OK. I suppose setting `worker-count: 1` can also restrict the entire processing to a single worker node?

yes, but the performance may not meet the requirements. and please note that if shard tables are at different mysql instances, it will be processed by different DM-workers, limiting worker-count will not help

**Comment by kennytm** [comment]
well if the DELETE and INSERT are on different MySQL instances they can't be in the same transaction anyway (assuming not using XA) so that's fine for this use case 🤣

**Comment by wxiaoke** [comment]
> This is not supported. Even if we use downstream table name in causality key, it can't handle the case where different shard tables are bind to different DM workers in different nodes. I think you can try to use `routes.${route-rule}.extract-table/extract-schema/extract-source` to extract information to extra columns, and make a composite primary key with these columns.
> 
> https://docs.pingcap.com/tidb/stable/task-configuration-file-full

If upstream is only one instance, Will we resolve this issue, Or just use the workaround?

**Comment by lance6716** [comment]
@wxiaoke I think it's not a workaround, it's the correct way to replicate data when merging shard tables. In other words, currently we have no plan to fix this issue.
